internal/volume/volumes_information.go

```diff
@@ -551,95 +550,124 @@ func (v *BackupVolumesInformation) generateVolumeInfoFromPVB() {
 	v.volumeInfos = append(v.volumeInfos, tmpVolumeInfos...)
 }
 
-// generateVolumeInfoFromDataUpload generate BackupVolumeInfo for DataUpload.
-func (v *BackupVolumesInformation) generateVolumeInfoFromDataUpload() {
-	if !features.IsEnabled(velerov1api.CSIFeatureFlag) {
-		v.logger.Debug("Skip generating BackupVolumeInfo when the CSI feature is disabled.")
-		return
+func (v *BackupVolumesInformation) getVolumeSnapshotClasses(
+	classes *snapshotv1api.VolumeSnapshotClassList,
+) (*snapshotv1api.VolumeSnapshotClassList, error) {
+	if classes != nil {
+		return classes, nil
 	}
 
-	tmpVolumeInfos := make([]*BackupVolumeInfo, 0)
 	vsClassList := new(snapshotv1api.VolumeSnapshotClassList)
 	if err := v.crClient.List(context.TODO(), vsClassList); err != nil {
-		v.logger.WithError(err).Errorf("cannot list VolumeSnapshotClass %s", err.Error())
+		v.logger.Warnf("Cannot list VolumeSnapshotClass with error %s.", err.Error())
+		return nil, err
+	}
+
+	return vsClassList, nil
+}
+
+// generateVolumeInfoFromDataUpload generate BackupVolumeInfo for DataUpload.
+func (v *BackupVolumesInformation) generateVolumeInfoFromDataUpload() {
+	if !features.IsEnabled(velerov1api.CSIFeatureFlag) {
+		v.logger.Debug("Skip generating BackupVolumeInfo when the CSI feature is disabled.")
 		return
 	}
 
+	// Retrieve the operations containing DataUpload.
+	duOperationMap := make(map[kbclient.ObjectKey]*itemoperation.BackupOperation)
 	for _, operation := range v.BackupOperations {
 		if operation.Spec.ResourceIdentifier.GroupResource.String() == kuberesource.PersistentVolumeClaims.String() {
-			var duIdentifier velero.ResourceIdentifier
-
 			for _, identifier := range operation.Spec.PostOperationItems {
 				if identifier.GroupResource.String() == "datauploads.velero.io" {
-					duIdentifier = identifier
+					duOperationMap[kbclient.ObjectKey{
+						Namespace: identifier.Namespace,
+						Name:      identifier.Name,
+					}] = operation
+
+					break
 				}
 			}
-			if duIdentifier.Empty() {
-				v.logger.Warnf("cannot find DataUpload for PVC %s/%s backup async operation",
-					operation.Spec.ResourceIdentifier.Namespace, operation.Spec.ResourceIdentifier.Name)
-				continue
-			}
+		}
+	}
+
+	var vsClassList *snapshotv1api.VolumeSnapshotClassList
+	if len(duOperationMap) > 0 {
+		var err error
+		vsClassList, err = v.getVolumeSnapshotClasses(vsClassList)
+		if err != nil {
+			return
+		}
+	} else {
+		// No DataUpload is found. Return early.
+		return
+	}
```

Comment:
If we remove parameter like lyndon said, this can just be like so. vsClassList will always be nil so parameter doesn't do anything.
```suggestion
	var vsClassList *snapshotv1api.VolumeSnapshotClassList
	if len(duOperationMap) > 0 {
		var err error
		vsClassList, err = v.getVolumeSnapshotClasses()
		if err != nil {
			return
		}
	} else {
		// No DataUpload is found. Return early.
		return
	}
```

Reply:
Modified.